src/unlock/signer.rs

```diff
@@ -608,46 +637,80 @@ impl OmniLockScriptSigner {
         let message = generate_message(&tx_new, script_group, zero_lock)?;
         let message = convert_keccak256_hash(message.as_ref());
 
-        let signature = self.signer.sign(
-            self.config.id().auth_content().as_ref(),
-            message.as_ref(),
-            true,
-            tx,
-        )?;
+        let signature = self
+            .signer
+            .sign(id.auth_content().as_ref(), message.as_ref(), true, tx)?;
 
-        let lock = OmniLockConfig::build_witness_lock(signature);
         // Put signature into witness
         let witness_data = witnesses[witness_idx].raw_data();
         let mut current_witness: WitnessArgs = if witness_data.is_empty() {
             WitnessArgs::default()
         } else {
             WitnessArgs::from_slice(witness_data.as_ref())?
         };
+
+        let lock = Self::build_witness_lock(current_witness.lock(), signature)?;
         current_witness = current_witness.as_builder().lock(Some(lock).pack()).build();
         witnesses[witness_idx] = current_witness.as_bytes().pack();
         Ok(tx.as_advanced_builder().set_witnesses(witnesses).build())
     }
+
+    /// Build proper witness lock
+    pub fn build_witness_lock(
+        orig_lock: BytesOpt,
+        signature: Bytes,
+    ) -> Result<Bytes, ScriptSignError> {
+        let lock_field = orig_lock.to_opt().map(|data| data.raw_data());
+        let omnilock_witnesslock = if let Some(lock_field) = lock_field {
+            OmniLockWitnessLock::from_slice(lock_field.as_ref())?
+        } else {
+            OmniLockWitnessLock::default()
+        };
+
+        Ok(omnilock_witnesslock
+            .as_builder()
+            .signature(Some(signature).pack())
+            .build()
+            .as_bytes())
+    }
 }
 
 impl ScriptSigner for OmniLockScriptSigner {
     fn match_args(&self, args: &[u8]) -> bool {
-        if !(args.len() == 22 && self.config.id().flag() as u8 == args[0]) {
+        if args.len() != self.config.get_args_len() {
+            return false;
+        }
+        if self.config.is_multisig() {
+            return self.config.id().auth_content().as_ref() == &args[1..21]
+                && self
+                    .config
+                    .multisig_config()
+                    .unwrap()
+                    .sighash_addresses
+                    .iter()
+                    .any(|id| self.signer.match_id(id.as_bytes()));
+        }
+        if self.config.omni_lock_flags().contains(OmniLockFlags::ADMIN) {
```

Comment:
Already checked in `match_rc_type_id()`, remove this line.

Reply:
if it's not admin mode, should check the normal mode.

Reply:
should remove the code of match_rc_type_id() here and remove the function match_rc_type_id